src/transformers/generation/logits_process.py

```diff
@@ -1102,3 +1102,119 @@ def __call__(self, input_ids, scores):
+
+
+    Examples:
+
+    ```python
+    >>> # base prompt emphasis example
+    >>> from transformers import AutoTokenizer, AutoModelForCausalLM
+
+    >>> model = AutoModelForCausalLM.from_pretrained("gpt2")
+    >>> tokenizer = AutoTokenizer.from_pretrained("gpt2")
+    >>> inputs = tokenizer(["Today, a dragon flew over Paris, France,"], return_tensors="pt")
+    >>> summary_ids = model.generate(inputs["input_ids"], guidance_scale=1.5)
+    >>> print(tokenizer.batch_decode(summary_ids, skip_special_tokens=True)[0])
+    The dragon flew over Paris, France, landing in Lyon, a city of a few million. Dragon-flying was a new form of
+    transport, and the dragon was the first in Europe.
+
+    >>> # with a negative prompt
+    >>> neg_inputs = tokenizer(["A very happy event happened,"], return_tensors="pt")
+    >>> summary_ids = model.generate(
+    ...     inputs["input_ids"], guidance_scale=5, cfg_rescale=0.3, negative_prompt=neg_inputs["input_ids"]
+    ... )
+    >>> print(tokenizer.batch_decode(summary_ids, skip_special_tokens=True)[0])
+    The dragon flew over Paris, France, crashing into Notre Dame Cathedral in the French capital killing at least 127
+    people and injuring more than 350.
+    ```
+    """
+
+    def __init__(
+        self,
+        guidance_scale,
+        unconditional_ids,
+        model,
+        rescale_factor=1.0,
+        use_cache=True,
+        unconditional_attention_mask=None,
+    ):
+        self.guidance_scale = guidance_scale
+        self.model = model
+        self.negative_context = {
+            "input_ids": unconditional_ids,
+            "attention_mask": unconditional_attention_mask or torch.ones_like(unconditional_ids, dtype=torch.long),
+            "use_cache": use_cache,
+            "past_key_values": None,
+            "first_pass": True,
+        }
+        self.rescale_factor = rescale_factor
+
+    def neg_logits(self, input_ids):
+        ctx = self.negative_context
+
+        if "first_pass" in ctx:
+            input_ids = ctx["input_ids"]
+            attention_mask = ctx["attention_mask"]
+            del ctx["first_pass"]
+        else:
+            attention_mask = torch.cat(
+                [ctx["attention_mask"], torch.ones_like(input_ids[:, -1:], dtype=torch.long)], dim=1
+            )
+            if not ctx["use_cache"]:
+                input_ids = torch.cat([ctx["input_ids"], input_ids[:, -1:]], dim=1)
+            else:
+                input_ids = input_ids[:, -1:]
+            ctx["input_ids"] = input_ids
+            ctx["attention_mask"] = attention_mask
+
+        out = self.model(
+            input_ids,
+            attention_mask=attention_mask,
+            use_cache=ctx["use_cache"],
+            past_key_values=ctx["past_key_values"],
+        )
+        ctx["past_key_values"] = out.get("past_key_values", None)
+
+        return out.logits
+
+    def __call__(self, input_ids, scores):
+        scores = torch.nn.functional.log_softmax(scores, dim=-1)
```

Comment:
I really like the idea proposed here and the quick HF implementation. However, I think the mathematics could be tightened up a bit here:
1. There is no need to call this log_softmax, nor the one on line 1215 -- calling logsoftmax only shifts the scores by a constant and that constant will cancel out in the final logsoftmax (line 1217)
2. I find the `rescale_factor` confusing and unnecessary. It's enough to have one hyperparameter to manipulate and the rescale factor can be absorbed into the `guidance_scale`. That is, unless I'm mistaken, `guidance_scale = \gamma` and `rescale_factor = \alpha` is equivalent to `guidance_scale = \alpha\gamma - \alpha + 1` and `rescale_factor = 1`.

Reply:
1. Indeed. We started thinking that those log_softmax were redundant, yes they do shift only by a constant. However, _this constant might not be the same_ in the conditional and unconditional branch, so those log_softmax sets them on common ground, resulting in higher performances. We actually started our experiment without those.
2. We tried to stick to [CFG Rescale](https://arxiv.org/abs/2305.08891) as close as possible.

Reply:
1. Hm, it's interesting that you would be seeing such performance differences, given that the calculations should be giving identical results. Indeed, the constants in the conditional and unconditional branches are different but they add up to a third constant (with value `-torch.logsumexp(logits) + self.guidance_scale * (torch.logsumexp(logits) - torch.logsumexp(scores))`) and that constant cancels in the final logsoftmax in line 1217. Would you attribute the performance differences you were seeing to numerical instabilities? 
2. You'd have to excuse me, I'm not really familiar with diffusion models, so maybe I'm missing something here. If I'm looking at the correct equation in the paper you linked (equation 16 on page 4?), then the rescale factor $\phi$ there seems to serve a different purpose -- it does not interpolate between $x_{pos}$ and $x_{cfg}$, instead it interpolates between $x_{cfg}$ and a rescaled version of $x_{cfg}$. The factor by which $x_{cfg}$ is rescaled appears to be data-dependent (a ratio of standard deviations) and so that doesn't allow for $\phi$ to be absorbed into other hyperparameters. This is not the case in the setting here though.

Reply:
Yes it's the correct equation. And... oh my. You're right. I misread the equation. It's totally on me. I would have bet it was interpolating back between x_pos and x_rescaled.

So, My current implementation works, that's for sure. It was battle tested with the GPT4All experiment. But maybe the correct interpretation of CFG rescale in this paper works better. Good catch.

Reply:
Thanks for looking into this. I get the interpretation with the "normalisation" and I think your analogy makes sense. But there is another point where things diverge and I think it has to do with the specific huggingface implementation. Namely, the output of `logits_processor` will subsequently undergo another `softmax` before the token is sampled (e.g. [here](https://github.com/huggingface/transformers/blob/v4.30.0/src/transformers/generation/utils.py?fbclid=IwAR3a4zhzIkQusT-Z796-160BuNmpqDJyNSfRauTUxqJn6IkworOCDPBcc2Q#L2654)). And this subsequent softmax does allow you to absorb the rescale_factor into the guidance_scale: I wrote a little test to confirm this: https://colab.research.google.com/drive/1ShO9q5UYjDApz6AlEY7-tiftG-CzY-5d?usp=sharing.

Out of curiousity, what values of the rescale_factor did you use in your experiments?